frontend/packages/ceph-storage-plugin/src/components/ocs-storage-class-form/ocs-storage-class-form.tsx

```diff
@@ -306,57 +311,132 @@ export const StorageClassEncryption: React.FC<ProvisionerProps> = ({
 const ExistingKMSDropDown: React.FC<ExistingKMSDropDownProps> = ({
   csiConfigMap,
   serviceName,
+  kmsProvider,
+  infraType,
   setEncryptionId,
 }) => {
   const { t } = useTranslation();
+  const isOdf = useFlag(ODF_MODEL_FLAG);
 
-  const [isOpen, setOpen] = React.useState(false);
+  const [isProviderOpen, setProviderOpen] = React.useState(false);
+  const [isServiceOpen, setServiceOpen] = React.useState(false);
+  const [provider, setProvider] = React.useState<string>(kmsProvider);
+  const [kmsServiceDropdownItems, setKmsServiceDropdownItems] = React.useState<JSX.Element[]>([]);
 
-  const handleDropdownChange = (e: React.KeyboardEvent<HTMLInputElement>) =>
-    setEncryptionId(e.currentTarget.id);
-  const kmsDropdownItems = _.map(csiConfigMap?.data, (connectionDetails, connectionName) => {
-    const kmsData = JSON.parse(connectionDetails);
-    return (
-      <DropdownItem
-        key={connectionName}
-        component="button"
-        id={connectionName}
-        data-test={connectionName}
-        onClick={handleDropdownChange}
-        description={kmsData?.VAULT_ADDR}
-      >
-        {connectionName}
-      </DropdownItem>
+  const handleProviderDropdownChange = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    setProvider(e.currentTarget.id);
+    setEncryptionId('');
+  };
+
+  const kmsProviderDropdownItems = _.reduce(
+    UISupportedProviders,
+    (res, providerDetails, providerName) => {
+      if (
+        !UISupportedProviders[providerName].allowedPlatforms ||
+        UISupportedProviders[providerName]?.allowedPlatforms.includes(infraType)
+      )
+        res.push(
+          <DropdownItem
+            key={providerDetails.group}
+            component="button"
+            id={providerName}
+            data-test={providerDetails.group}
+            onClick={handleProviderDropdownChange}
+          >
+            {providerDetails.group}
+          </DropdownItem>,
+        );
+      return res;
+    },
+    [],
+  );
+
+  React.useEffect(() => {
+    const handleServiceDropdownChange = (e: React.KeyboardEvent<HTMLInputElement>) =>
+      setEncryptionId(e.currentTarget.id);
+    setKmsServiceDropdownItems(
+      _.reduce(
+        csiConfigMap?.data,
+        (res, connectionDetails, connectionName) => {
+          try {
+            // removing any object having syntax error
+            // or, which are not supported by UI.
+            const kmsData = JSON.parse(connectionDetails);
+            if (UISupportedProviders[provider].supported.includes(kmsData?.[KMS_PROVIDER])) {
+              res.push(
+                <DropdownItem
+                  key={connectionName}
+                  component="button"
+                  id={connectionName}
+                  data-test={connectionName}
+                  onClick={handleServiceDropdownChange}
+                  description={kmsData?.[DescriptionKey[kmsData?.[KMS_PROVIDER]]]}
```

Comment:
Are you using vault and ibmKms as kms provider also, Also vaulttokens and ibmkeyprotect as kms providers?  this is confusing.

Reply:
Ideally vault and ibmroks should be the kms providers (that is what I am using everywhere in the code)...
but, we have a key "KMS_PROVIDER" in the configMap whose value can be "vaulttokens", "ibmkeyprotect", "vaulttenantsa" etc. (which means provider+auth-method)... that is why this is causing confusion.

That is I suppose, key naming issue from the backend side. configMap's name should have been "KMS_IMPLEMENTATION" or something similar.